src/DotLiquid/Tokenizer.cs

```diff
@@ -124,32 +194,100 @@ private static void ReadToEnd(StringBuilder sb, DotLiquid.Util.CharEnumerator ma
                     switch (nextChar)
                     {
                         case '\'':
-                            isInQuotes = !isInQuotes;
-                            searchChars = isInQuotes ? SearchSingleQuoteEnd : initialSearchChars;
-                            break;
                         case '"':
-                            isInQuotes = !isInQuotes;
-                            searchChars = isInQuotes ? SearchDoubleQuoteEnd : initialSearchChars;
+                            ReadToChar(sb, markupEnumerator, nextChar);
                             break;
                         case '}':
                         case '%':
                             if (markupEnumerator.Remaining > 0 && markupEnumerator.Next == '}')
                             {
+                                var previousCharIsWhitespaceControl = syntaxCompatibilityLevel >= SyntaxCompatibility.DotLiquid22 && markupEnumerator.Previous == '-';
                                 markupEnumerator.AppendNext(sb);
-                                return;
+                                if (previousCharIsWhitespaceControl)
+                                {
+                                    // Remove hyphen from token
+                                    sb.Remove(sb.Length - 3, 1);
+
+                                    // Trim trailing whitespace by skipping ahead beyond the tag end
+                                    while (markupEnumerator.Remaining > 0)
+                                    {
+                                        if (((uint)markupEnumerator.Next - '\t') <= 5 || markupEnumerator.Next == ' ')
```

Comment:
I based this on MS stating they do this for RegEx character classes in .NET 5.0. I guess it is more performant and technically we should change the other trim to potentially use this. For now, this will do. Once it is accepted we can do performance testing to see if there is any value to changing.